### PR TITLE
Enable kustomize in kubectl

### PR DESCRIPTION
**Pull request by Liujingfang1**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This PR is the implementation of [KEP](https://github.com/kubernetes/community/pull/2910) to enable kustomize in kubectl. 

When `-f <dir>` is passed to a kubectl command, it will look for a `kustomization.yaml` file. If  `kustomization.yaml` is found, a kustomize build will be run to get the list of expanded resources. This list of resources is then passed to kubectl commands. If there is no `kustomization.yaml` in the directory, kubectl will behave the same as current.

To apply a kustomization directory
```
kubectl apply -f <dir>
```
To get resources of a kustomization directory applied to a cluster
```
kubectl get -f <dir>
```
To delete a kustomization directory applied to a cluster
```
kubectl delete -f <dir>
```

**Special notes for your reviewer**:

This PR contains 6 commits. 
- The first three of them is to vendor kustomize.
- The 4th commit is the change in resource builder to use Kustomization when it it enabled. 
- The 5th one adds some unit test for Builder.
- The 6th commit is to enable kustomization in kubectl commands.

Kubectl will have kustomization enabled by default.
Other `cli-runtime` consumers may choose if they want to enable or not kustomization by a Boolean variable.

**Does this PR introduce a user-facing change?**: 
NONE

```release-note
Enable customize in kubectl: kubectl will be able to recognize directories with kustomization.YAML
```

### CODE REVIEW

File: staging/src/k8s.io/cli-runtime/pkg/genericclioptions/resource/visitor.go
Comment (smarterclayton) [inline staging/src/k8s.io/cli-runtime/pkg/genericclioptions/resource/visitor.go]: I don't think this is backwards compatible.  Anyone using `kubectl create -f DIR` will see something different happen after this change lands, which means existing CLI workflows could break.

**Review by smarterclayton — CHANGES_REQUESTED** [review]

File: staging/src/k8s.io/cli-runtime/pkg/genericclioptions/resource/visitor.go
Comment (justinsb) [reply staging/src/k8s.io/cli-runtime/pkg/genericclioptions/resource/visitor.go]: I just tested the behaviour.  `kubectl create` and `kubectl apply` both fail with a kustomization.yaml present:

```
error: error validating ".../k8s.io/examples/guestbook-go/kustomization.yaml": error validating data: [apiVersion not set, kind not set]; if you choose to ignore these errors, turn validation off with --validate=false
```

It looks like we have a separate issue, which is that we don't validate all files before starting to apply, which is contrary to what I would have expected.

But I don't think users can be using `kubectl create -f DIR` or `kubectl apply -f DIR` today with a dir containing kustomization.yaml.

Comment (Liujingfang1) [reply staging/src/k8s.io/cli-runtime/pkg/genericclioptions/resource/visitor.go]: @smarterclayton The existing kubectl doesn't work with directories with a kustomization.yaml as @justinsb explained. With this PR, kubectl will be able to recognize a directory with a kustomization.yaml. For any directories without kustomization.yaml, there is no change in kubectl's behavior.

Comment (smarterclayton) [reply staging/src/k8s.io/cli-runtime/pkg/genericclioptions/resource/visitor.go]: If someone has validate=false off, what happens?  If it also fails, then my primary concern is addressed.

Comment (Liujingfang1) [reply staging/src/k8s.io/cli-runtime/pkg/genericclioptions/resource/visitor.go]: With validate=false, it fails with similar error
```
error: unable to decode "kustomization.yaml": Object 'Kind' is missing in `<truncated>`
```

**Comment by neolit123** [comment]
@Liujingfang1 thank you for working on this, please add release note explaining the change.
/priority important-longterm

golang makes including third party code way too easy.

my problem with kustomize is that it vendors the whole AWS-SDK.
kubernetes already vendors it for other purposes, but i don't think in the long run kubernetes should vendor cloud provider SDKs, instead cloud provider implementers should vendor kubernetes interface packages and then cloud provider SDKs.

for kustomize i proposed that the AWS-SDK usage to be made pluggable - i.e. a plugin.
if this is not done kubectl will now depend on the AWS-SDK.

@kubernetes/sig-architecture-pr-reviews

**Comment by Liujingfang1** [comment]
@neolit123 Thank you for reviewing the change. `aws-sdk` was added as a transitive dependency of kustomize. The dependency chain is: Kustomize  -> go-getter -> aws-sdk. It is used to read files from remote locations like GitHub repos.

We're going to remove the dependency of `go-getter` from Kustomize. https://github.com/kubernetes-sigs/kustomize/issues/559 With that, we are going to update this PR. Then aws-sdk will not be included.

**Comment by neolit123** [comment]
thanks for the clarification @Liujingfang1

**Comment by neolit123** [comment]
^ cc @BenTheElder

**Comment by monopole** [comment]
@Liujingfang1 feel free to rebase; go-getter removed in kubernetes-sigs/kustomize#579

**Comment by Liujingfang1** [comment]
Rebased for both kubernetes and kustomize changes.  All tests passed. PTAL

**Comment by Liujingfang1** [comment]
@cblecker Can you take a look at the dep change? Only `sigs.k8s.io/kustomize/pkg` is added. Thank you.

**Comment by Liujingfang1** [comment]
@lavalamp Can you take a look at the staging change? The change is that `sigs.k8s.io/kustomize/pkg` is added to `cli-runtime` with its transitive dependencies. Thank you.

**Comment by cblecker** [comment]
No concerns with the Godep import changes. Can't /approve yet, as I can't speak to the import restriction changes (those should be reviewed by sig-cli).

File: staging/src/k8s.io/cli-runtime/pkg/genericclioptions/resource/visitor.go
Comment (pwittrock) [inline staging/src/k8s.io/cli-runtime/pkg/genericclioptions/resource/visitor.go]: why not `filepath.Base(path) == "kustomization.yaml"`

Comment (liggitt) [reply staging/src/k8s.io/cli-runtime/pkg/genericclioptions/resource/visitor.go]: I thought the intent of cli-runtime was to be generic function for things consuming resource builder functionality. Hard-coding this seems extremely specific and not very scalable.

Can we pull this out to a package var or registration method that kubectl could inject special handling for `kustomization.yaml` into, rather than making all cli-runtime consumers pick up `kustomize`?

Comment (juanvallejo) [reply staging/src/k8s.io/cli-runtime/pkg/genericclioptions/resource/visitor.go]: It sounds to me like there are two possible approaches to the kustomize implementation:

1. kustomize remains unique to kubectl, it lives in kubectl, and any extra behavior it needs to introduce to the resource-builder (or any other generic utilities used to consume resources) is done from a "initialization" function within kubectl.

2. kustomize becomes a part of the overall cli ecosystem, we keep it hardcoded within genericclioptions, and `kubectl` along with third-party clients and plugins all inherit this new capability of handling "kustomize" directories

I would be fine with either of these two approaches, however I do see value in propagating this behavior to consumers of genericclioptions, as it would maintain this similar behavior with kubectl.

cc @soltysh @smarterclayton @deads2k for additional thoughts on this

Comment (Liujingfang1) [reply staging/src/k8s.io/cli-runtime/pkg/genericclioptions/resource/visitor.go]: @pwittrock  I'll update the statement. @smarterclayton  We can change the hard coded `kustomization.yaml` by a var or registration method. That will require the change from the kustomize side. Can we do it in a following PR?

@juanvallejo I prefer it becomes a part of the overall cli ecosystem so that other clients can also use kustomize by default.

Comment (soltysh) [reply staging/src/k8s.io/cli-runtime/pkg/genericclioptions/resource/visitor.go]: So Juan and I discussed this in depth and here's our proposed solution. We picked a middle-ground solution from what he proposed in the previous comment. This means that cliruntime will have extension point (like Jordan proposes) where an author would be able to include additional functionality (eg. `cliruntime.Addons["kustomization.yaml"] = kustomizationHandler()`). Kustomize would be such an extensions, our first one. It would entirely live inside its current repo (https://github.com/kubernetes-sigs/kustomize). 
This way kubectl (as its first consumer) would vendor both cliruntime and kustomize and initilize kustomize as an addon for cliruntime. This would allow few things:

1. keep cliruntime generic and very minimal
2. provide extensions points for cliruntime (builder to start with)
3. give cliruntime consumers bigger flexibility over what they're getting.

@pwittrock @liggitt @juanvallejo @deads2k @smarterclayton what do you think
@Liujingfang1 I'm sorry we're changing the design one more time :relaxed:

**Review by soltysh — CHANGES_REQUESTED** [review]
/hold 
until we agree on the final shape

File: staging/src/k8s.io/cli-runtime/pkg/genericclioptions/resource/visitor.go
Comment (pwittrock) [reply staging/src/k8s.io/cli-runtime/pkg/genericclioptions/resource/visitor.go]: @soltysh Didn't see your comment, responses inlined.

> 1. keep cliruntime generic and very minimal

This also makes the default behavior for clients inconsistent with the default behavior for kubectl commands such as apply, get, describe.  May be it would be best to have a simple and modular architecture by making all of the Resource Builder targets be plugins.

> 2. provide extensions points for cliruntime (builder to start with)

I'd like to see more discussion about this and what the interfaces might look like before committing to it.

> give cliruntime consumers bigger flexibility over what they're getting.

This can be done more simply by making it opt-out

Comment (monopole) [reply staging/src/k8s.io/cli-runtime/pkg/genericclioptions/resource/visitor.go]: For context, from the cli-runtime readme:

> [Purpose](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/cli-runtime): 
> This library is a shared dependency for clients to work with Kubernetes API
> infrastructure which allows to maintain kubectl compatible behavior.
> Its first consumer is k8s.io/kubectl.

To be useful in a k8s context, cli-runtime must be cognizant of API [guidelines](https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md) and recognize (depend on) things like `TypeMeta` and `ObjectMeta`, and be able to wrangle YAML/JSON, rest, patching, etc, even as it avoids direct dependence on the API types.

kustomize was built with the same goals and restrictions.  It was developed as a standalone project (outside k/k) as an expediency to release faster, get feedback faster, prove utility and to avoid making the kubectl extraction project more difficult.  It was always intended to become part of the builder, to enhance the CLI’s ability to manipulate k8s-style objects in multiple commands and contexts. 

It’s a cross-cutting enhancement to manipulate k8s YAML meant to collectively address [old kubectl bugs mentioned in the KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cli/0008-kustomize.md#long-standing-issues).  We should fix these issues for all runtime consumers, not just kubectl users.

So please accept this PR (more or less as is, modulo how the file name is handled), and if there's some outcry from CLI users who don't want these fixes/features, we can factor it out of the runtime then.

**Comment by pwittrock** [comment]
Lets finalize decisions on this during the contributor summit or kubectl meeting next week.

**Comment by monopole** [comment]
looks to me like the new enableKustomization option addresses the concerns expressed above.
/lgtm

**Comment by thockin** [comment]
Will this live in sigs.k8s.io forever or will all of this code eventually find itself together?

/approve

**Comment by Liujingfang1** [comment]
As discussed offline, we can remove the opt out. I added a commit for that. @pwittrock PTAL

File: staging/src/k8s.io/cli-runtime/pkg/genericclioptions/resource/visitor.go
Comment (pwittrock) [inline staging/src/k8s.io/cli-runtime/pkg/genericclioptions/resource/visitor.go]: This will be changed to check the GVK of the kustomization file in an immediate follow up.

File: staging/src/k8s.io/cli-runtime/pkg/genericclioptions/resource/visitor.go
Comment (pwittrock) [inline staging/src/k8s.io/cli-runtime/pkg/genericclioptions/resource/visitor.go]: In an immediate follow up this should be a library that takes options.

**Comment by pwittrock** [comment]
Follow up issues tracked here:
https://github.com/kubernetes/kubectl/issues/570

**Comment by pwittrock** [comment]
@soltysh removing hold.  @JiangtianLi is working on a follow up.

**Comment by anguslees** [comment]
I'm only just learning about kustomize, but I am a bit alarmed about what I'm reading and the implications for this PR.  With this PR, can I just pwn the world by putting a malicious kustomize.yaml in a popular manifest examples site somewhere?

Consider:
```yaml
secretGenerator:
- name: allyourbase
  commands:
    # or any other malicious command
    foo: "echo backdoorkey >> $HOME/.ssh/authorized_keys"
```

In particular, with this PR, I think `kubectl apply -f http://that/repo" suddenly becomes able to modify the *local* machine, not just the target cluster, even with `--dry-run`.

**Comment by BenTheElder** [comment]
FWIW I think that particular issue will be covered by
https://github.com/kubernetes-sigs/kustomize/pull/683.
This PR as-is was merged early in the current release cycle so it will not
be in any releases (not even an alpha) for a bit.

**Comment by pwittrock** [comment]
Discussion of changes to the UX integration were raised and are being talked through in kubernetes/kubectl#570. The outcome of those discussions will be folded into the KEP before reintegrating: kubernetes/enhancements#684

Additionally there were some security concerns raised that require changes, such as limiting process callouts: kubernetes-sigs/kustomize#683. The new capabilities added by kustomize will be reviewed from a security perspective prior to reintegration.

PR to revert: kubernetes/kubernetes#72805

**Pull request merged**
